### DISCUSSION TITLE
Ammo and Turret

### DISCUSSION THREAD

**Discussion by kobango** [FakeFishGames/Barotrauma#12532] [Code questions and modding support]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

I try to make mod to place turret in game without sub edytor.
As the linking is not possible in game i was forced to make turrent self sustaining (Ammo on item)
It's work pretty good with one exception, boot are confused and trying to load ammo to non existing rack, and more i cannot see ammo on gui. I trying to replicate ItemContainer but it's not work. So... a set LaunchWithoutProjectile to true, but boot's still search for ammo and still are unable to fire. I'm was thinking on make this full auto targeting but it's to OP when its use orginal ammunition.

Recreating loader is not problem but linking it is.

I humbly ask for any help or ideas.
![image](https://github.com/Regalis11/Barotrauma/assets/33550249/73ada10e-64ad-42b1-a1d3-7e823e171d99)
![image](https://github.com/Regalis11/Barotrauma/assets/33550249/6d88cd8e-bedd-4030-8edc-4e38057227cc)

### Reproduction steps

1. Add xml mentioned in important messeges section to game
2. Observe NPC asigned to turret

### Bug prevalence

Just once

### Single player or multiplayer?

Single player

### Version

v1.0.21.0

### Which operating system did you encounter this bug on?

Windows

### Relevant error messages and crash reports

```shell
<Item name="Mini chaingun" 
        description=""
        identifier="mini_chaingun" Tags="smallitem,turret" category="Machine,Weapon" interactthroughwalls="true" Scale="0.3" interactdistance="10" spritecolor="1.0,1.0,1.0,1.0" focusonselected="true" offsetonselected="750" linkable="true" allowedlinks="chaingunequipment" impactsoundtag="impact_metal_light">
	<Price baseprice="1500" minleveldifficulty="15">
		<Price storeidentifier="merchantoutpost" />
		<Price storeidentifier="merchantcity" sold="false"/>
		<Price storeidentifier="merchantresearch" multiplier="1.25" />
		<Price storeidentifier="merchantmilitary" />
		<Price storeidentifier="merchantmine" />
		<Price storeidentifier="merchantengineering" multiplier="0.9" />
	</Price>
	<PreferredContainer primary="engcab"/>
	<Deconstruct time="10">
		<Item identifier="plastic" />
		<Item identifier="copper" />
		<Item identifier="silicon" />
	</Deconstruct>
	<Fabricate suitablefabricators="fabricator" requiredtime="10">
		<RequiredSkill identifier="electrical" level="30" />
		<RequiredItem identifier="fpgacircuit" />
		<RequiredItem identifier="titaniumaluminiumalloy" amount="3" />
		<RequiredItem identifier="steel" amount="2" />
	</Fabricate>
	<InventoryIcon texture="Content/Items/Weapons/Loaders.png" sourcerect="0,0,256,256" />
	<Sprite texture="Content/Items/Weapons/Loaders.png" depth="0.01" sourcerect="0,0,256,256" canflipy="false" />
    <MinimapIcon name="Command_Weapons_Chaingun" texture="Content/UI/CommandUIAtlas.png" sourcerect="384,896,128,128" />
	  
    <StaticBody width="80" radius="80" />
	<Body width="256" height="256" density="15" />
	<Holdable selectkey="Select" pickkey="Action" slots="Any,RightHand,LeftHand" msg="ItemMsgDetachWrench" PickingTime="5.0" aimpos="65,-10" handle1="0,0" attachable="true" aimable="true">
		<RequiredItem items="wrench" type="Equipped" />
	</Holdable>
    <Turret launchimpulse="100" spinningbarreldistance="45.0" firingrotationspeedmodifier="0.3" usefiringoffsetformuzzleflash="true" maxchargetime="1.0" canbeselected="false" firingoffset="-10,-370" characterusable="false" linkable="false" barrelpos="128,88" rotationlimits="180,360" powerconsumption="400.0" showchargeindicator="true" showprojectileindicator="true" recoildistance="50" reload="0.3" springstiffnesslowskill="2" springstiffnesshighskill="50" springdampinglowskill="0.5" springdampinghighskill="10" rotationspeedlowskill="1" rotationspeedhighskill="5" LaunchWithoutProjectile="true">
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot1.ogg" type="OnUse" range="10000" selectionmode="Random" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot2.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot3.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot4.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot5.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot6.ogg" type="OnUse" range="10000" />
      <WeaponIndicator texture="Content/UI/WeaponUI.png" sourcerect="575,914,66,32" origin="0.227, 0.531" />
      <RailSprite texture="%ModDir%/Loaders.png" depth="0.011" sourcerect="256,0,212,512" origin="0.425, 0.875" />
      <SpinningBarrelSprite spriteamount="3" texture="%ModDir%/Loaders.png" depth="0.013" sourcerect="468,0,46,396" origin="0.1, 1.4" scale="1.0" />
      <Crosshair texture="Content/Items/Weapons/Crosshairs.png" sourcerect="256,0,256,256" />
      <CrosshairPointer texture="Content/Items/Weapons/Crosshairs.png" sourcerect="0,0,256,256" />
      <MoveSound file="Content/Items/Weapons/RailgunLoop.ogg" />
      <StartMoveSound file="Content/Items/Weapons/RailgunStart.ogg" />
      <EndMoveSound file="Content/Items/Weapons/RailgunStop.ogg" />
      <ChargeSound file="Content/Items/Weapons/WEAPONS_chargeUp.ogg" range="10000" />
      <RequiredSkill identifier="weapons" level="50" />
      <LightComponent LightColor="1.0,0.8,0.8,1.0" Flicker="0.0" range="2500" IsOn="true" drawbehindsubs="true" ignorecontinuoustoggle="true">
        <LightTexture texture="Content/Lights/lightcone.png" origin="0.0, 0.5" size="1.0,1.0" />
      </LightComponent>
      <ParticleEmitter particle="muzzleflashchaingun" particleamount="1" velocitymin="0" velocitymax="0" distancemin="-50" distancemax="-50" />
      <ParticleEmitter particle="muzzleflash" particleamount="1" velocitymin="0" velocitymax="0" scalemin="2" scalemax="3"/>
      <ParticleEmitter particle="swirlysmoke" particleamount="10" velocitymin="0" velocitymax="0" scalemin="2" scalemax="3" distancemin="-100" distancemax="50" />
      <StatusEffect type="OnUse" target="This">
        <Explosion range="1500.0" structuredamage="0" force="0.0" camerashake="8" flames="false" smoke="false" sparks="false" underwaterbubble="false" />
      </StatusEffect>
    </Turret>
	<ItemContainer hideitems="false" drawinventory="true" capacity="1" maxstacksize="1" slotsperrow="6" itempos="75,-270" iteminterval="0,0" itemrotation="0" canbeselected="true" msg="ItemMsgInteractSelect" containedspritedepth="0.79">
		<GuiFrame relativesize="0.18,0.23" anchor="Center" minsize="140,170" maxsize="280,280" style="ItemUI" />
		<Containable items="chaingunammo" />
		<!-- when the chaingun is fired, it triggers this statuseffect, causing contained ammunition boxes to spawn new ammo -->
		<StatusEffect type="OnUse" target="Contained">
			<RequiredItem items="chaingunammo" type="Contained" />
			<Use />
		</StatusEffect>
	</ItemContainer>
    <aitarget maxsightrange="3000" maxsoundrange="8000" fadeouttime="5" />
    <ConnectionPanel selectkey="Action" canbeselected="true" msg="ItemMsgRewireScrewdriver" hudpriority="10">
      <GuiFrame relativesize="0.2,0.32" minsize="400,350" maxsize="480,420" anchor="Center" style="ConnectionPanel" />
      <RequiredItem items="screwdriver" type="Equipped" />
      <input name="power_in" displayname="connection.powerin" />
      <input name="position_in" displayname="connection.turretaimingin" />
      <input name="trigger_in" displayname="connection.turrettriggerin" />
      <input name="toggle_light" displayname="connection.togglelight"/>
      <input name="set_light" displayname="connection.setlight" />
    </ConnectionPanel>
    <SkillRequirementHint identifier="weapons" level="50" />
  </Item>
```

**Comment by Regalis11** [comment]
Thank you for the report!

This seems to be a request for modding support rather than a bug report, so I'll move this to the Discussions.

On a cursory look though, I think the issue here is how the turret tries to spawn ammo inside itself. It appears you have copied the status effect that makes the chaingun loader spawn more ammo inside the ammo box, but in this case, there is no ammo box. I think you need a status effect that spawns chaingun bolts inside the turret.

**Comment by kobango** [comment]
Thanks a lot for the replay. It means a lot to me.
I'm trying to set the status inside the tower but it doesn't work.
`<Turret launchimpulse="100" spinningbarreldistance="45.0" firingrotationspeedmodifier="0.3" usefiringoffsetformuzzleflash="true" maxchargetime="1.0" canbeselected="false" firingoffset="-10,-370" characterusable="false" linkable="false" barrelpos="128,88" rotationlimits="180,360" powerconsumption="400.0" showchargeindicator="true" showprojectileindicator="true" recoildistance="50" reload="0.3" springstiffnesslowskill="2" springstiffnesshighskill="50" springdampinglowskill="0.5" springdampinghighskill="10" rotationspeedlowskill="1" rotationspeedhighskill="5" LaunchWithoutProjectile="true">
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot1.ogg" type="OnUse" range="10000" selectionmode="Random" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot2.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot3.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot4.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot5.ogg" type="OnUse" range="10000" />
      <Sound file="Content/Items/Weapons/WEAPONS_chainGunShot6.ogg" type="OnUse" range="10000" />
      <WeaponIndicator texture="Content/UI/WeaponUI.png" sourcerect="575,914,66,32" origin="0.227, 0.531" />
      <RailSprite texture="%ModDir%/Loaders.png" depth="0.011" sourcerect="256,0,212,512" origin="0.425, 0.875" />
      <SpinningBarrelSprite spriteamount="3" texture="%ModDir%/Loaders.png" depth="0.013" sourcerect="468,0,46,396" origin="0.1, 1.4" scale="1.0" />
      <Crosshair texture="Content/Items/Weapons/Crosshairs.png" sourcerect="256,0,256,256" />
      <CrosshairPointer texture="Content/Items/Weapons/Crosshairs.png" sourcerect="0,0,256,256" />
      <MoveSound file="Content/Items/Weapons/RailgunLoop.ogg" />
      <StartMoveSound file="Content/Items/Weapons/RailgunStart.ogg" />
      <EndMoveSound file="Content/Items/Weapons/RailgunStop.ogg" />
      <ChargeSound file="Content/Items/Weapons/WEAPONS_chargeUp.ogg" range="10000" />
      <RequiredSkill identifier="weapons" level="50" />
      <LightComponent LightColor="1.0,0.8,0.8,1.0" Flicker="0.0" range="2500" IsOn="true" drawbehindsubs="true" ignorecontinuoustoggle="true">
        <LightTexture texture="Content/Lights/lightcone.png" origin="0.0, 0.5" size="1.0,1.0" />
      </LightComponent>
      <ParticleEmitter particle="muzzleflashchaingun" particleamount="1" velocitymin="0" velocitymax="0" distancemin="-50" distancemax="-50" />
      <ParticleEmitter particle="muzzleflash" particleamount="1" velocitymin="0" velocitymax="0" scalemin="2" scalemax="3"/>
      <ParticleEmitter particle="swirlysmoke" particleamount="10" velocitymin="0" velocitymax="0" scalemin="2" scalemax="3" distancemin="-100" distancemax="50" />
      <StatusEffect type="OnUse" target="This">
        <Explosion range="1500.0" structuredamage="0" force="0.0" camerashake="8" flames="false" smoke="false" sparks="false" underwaterbubble="false" />
      </StatusEffect>
	  <StatusEffect type="OnUse" target="Contained">
			<RequiredItem items="chaingunammo" type="Contained" />
			<Use />
	  </StatusEffect>
		
    </Turret>
	<ItemContainer hideitems="false" drawinventory="true" capacity="1" maxstacksize="1" slotsperrow="6" itempos="150,-170" iteminterval="0,0" itemrotation="0" canbeselected="true" msg="ItemMsgInteractSelect" containedspritedepth="0.79">
			<GuiFrame relativesize="0.18,0.23" anchor="Center" minsize="140,170" maxsize="280,280" style="ItemUI" />
			<Containable items="chaingunammo" />
			
			
	</ItemContainer>`
also I make this container inside (assuming that linking work on object inside item as well as on item) but it's also not make ammo gui visible.
I make sure that the showprojectileindicator="true" is on true.
I have no other ideas, for player's it's not that problem, for bot's it's curse. Only solution is to make turret full automate with autooperate, I only regret that auto operatre turret do not consume energy on shot, not require power to operate, i read that shoud be fixed so I have no choice but to wait.

I also notice the characterusable="false" in turret, but it's apear on raillgun as well so I'm not sure how it's work. Maybe it's prevents ammo from being used ( but ammo is being used correctly for some reason).

Thanks again and if anyone has any ideas please help.

**Comment by kobango** [comment]
`List<Item> availableAmmo = new List<Item>();
            foreach (MapEntity e in item.linkedTo)
            {
                if (!(e is Item linkedItem)) { continue; }
                var itemContainer = linkedItem.GetComponent<ItemContainer>();
                if (itemContainer == null) { continue; }
                availableAmmo.AddRange(itemContainer.Inventory.AllItems);
                for (int i = 0; i < itemContainer.Inventory.Capacity - itemContainer.Inventory.AllItems.Count(); i++)
                {
                    availableAmmo.Add(null);
                }
            }
`
I find it now in project, ammo is only searched via link
